src/value/owned.rs

```diff
@@ -253,51 +287,59 @@ impl IonStruct for Struct {
-        self.text_fields
-            .get(field_name.as_ref())?
-            .last()
-            .map(|(_s, v)| v)
-    }
-
-    fn get_all<'a, T: AsRef<str>>(
+        Box::new(self.fields())
+    }
+
+    fn get<A: AsSymbolRef>(&self, field_name: A) -> Option<&Self::Element> {
+        match field_name.as_symbol_ref().text() {
+            None => {
+                // Build a cheap, stack-allocated `Symbol` that represents unknown text
+                let symbol = Symbol::unknown_text();
+                // Use the unknown text symbol to look up matching field values
+                self.fields.get(&symbol)?.last()
+            }
+            Some(text) => {
+                // Otherwise, look it up by text
+                self.fields.get(text)?.last()
+            }
+        }
+    }
+
+    fn get_all<'a, A: AsSymbolRef>(
         &'a self,
-        field_name: T,
+        field_name: A,
     ) -> Box<dyn Iterator<Item = &'a Self::Element> + 'a> {
-        Box::new(
-            self.text_fields
-                .get(field_name.as_ref())
-                .into_iter()
-                .flat_map(|v| v.iter())
-                .map(|(_s, v)| v),
-        )
+        // TODO: This method has to box its return value, which is why we're not re-using it
+        //       for the implementation of the `get` method. Once we remove the `Box`
+        //       (via GATs or RPITIT?) we can have `get` call `get_all`.
+        let values = match field_name.as_symbol_ref().text() {
+            None => {
+                // Build a cheap, stack-allocated `Symbol` that represents unknown text
+                let symbol = Symbol::unknown_text();
+                self.fields.get(&symbol)
+            }
+            Some(text) => {
+                // Otherwise, look it up by text
+                self.fields.get(text)
+            }
+        };
+
+        match values {
+            Some(values) => Box::new(values.iter()),
+            None => Box::new(None.iter()), // An empty iterator
+        }
     }
 }
 
 impl PartialEq for Struct {
     fn eq(&self, other: &Self) -> bool {
         // check if both text_fields and no_text_fields have same length
-        self.text_fields.len() == other.text_fields.len() && self.no_text_fields.len() == other.no_text_fields.len()
-            // check if text_fields and no_text_fields are equal
+        self.len() == other.len()
             // we need to test equality in both directions for both text_fields and no_text_fields
             // A good example for this is annotated vs not annotated values in struct
             //  { a:4, a:4 } vs. { a:4, a:a::4 } // returns true
             //  { a:4, a:a::4 } vs. { a:4, a:4 } // returns false
-            && self.eq_text_fields(other) && other.eq_text_fields(self)
-            && self.eq_no_text_fields(other) && other.eq_no_text_fields(self)
+            && self.fields_eq(other) && other.fields_eq(self)
     }
 }
 
```

Comment:
We should probably add some tests for `len` and `is_empty` functions.

Reply:
Agreed!

Reply:
Added some unit tests for this.